### PR TITLE
PUBDEV-6438: Implement 2D partial dependency plots

### PR DESCRIPTION
**Pull request by wendycwong**
This PR completes JIRA: https://0xdata.atlassian.net/browse/PUBDEV-6438?filter=-1

I have completed the following:
1. Added Java backend to support 2d pdp;
2. Added Junit test to make sure both 1d and 2d pdp works;
3. Added python client side to support specifying 2d pdp.
4. Added pyunit test;
5. Added R client side support to allow 2d pdp
6. Added Runit test.
7. plot 3d graph instead of 2d for both R and Python

### CODE REVIEW

File: h2o-core/src/main/java/hex/PartialDependence.java [hunk 1]
```diff
@@ -94,151 +116,208 @@ private void checkSanityAndFillParams() {
...
+                  preds = _model_id.get().score(test, Key.make().toString(), _job, false);
+                  if (preds == null || preds.numRows() == 0) {  // this can happen if algo will not predict on rows with NAs
+                    meanResponse[which] = Double.NaN;
+                    stddevResponse[which] = Double.NaN;
+                    ;
+                    stdErrorOfTheMeanResponse[which] = Double.NaN;
+                    ;
+                  } else {
+                    if (_model_id.get()._output.nclasses() == 2) {
+                      if (_weight_column_index >= 0) { // calculated weighted statistics
```
Comment (honzasterba) [inline h2o-core/src/main/java/hex/PartialDependence.java]: i think we are here about four leves of nested blocks too deep

File: h2o-algos/src/test/java/hex/pdp/PartialDependenceTest.java [hunk 2]
```diff
@@ -121,6 +122,127 @@ assert equalTwoDimTables(partialDependence._partial_dependence_data[1], partialD
...
+      assertCorrec2Dpdp(fr, partialDependence._partial_dependence_data[2].getCellValues(), "AGE", "RACE", 
+              false, true, ageSplit, raceSplit, model, tot, tstats);  // check out the first 2d pdp betwee AGE and RACE
+      assertCorrec2Dpdp(fr, partialDependence._partial_dependence_data[3].getCellValues(), "AGE", "PSA",
+              false, false, ageSplit, psaSplit, model, tot, tstats);  // check out the first 2d pdp betwee AGE and PSA
+    } finally {
+      Scope.exit();
+    }
+  }
+  
+  public void assertCorrec2Dpdp(Frame fr, IcedWrapper[][] cellVs, String col, String col2,  boolean cat, boolean cat2,
```
[the remark this answers is not in the document]
Comment (wendycwong) [reply h2o-algos/src/test/java/hex/pdp/PartialDependenceTest.java]: Definitely a typo.  Fixed

File: h2o-algos/src/test/java/hex/pdp/PartialDependenceTest.java [hunk 3]
```diff
@@ -121,6 +122,127 @@ assert equalTwoDimTables(partialDependence._partial_dependence_data[1], partialD
...
+  public void assertCorrec2Dpdp(Frame fr, IcedWrapper[][] cellVs, String col, String col2,  boolean cat, boolean cat2, 
+                                double[] colVals, double[] col2Vals, GBMModel model, double tot, double[] tstats) {
+    for (int index=0; index < cellVs.length; index++) {
+      int counter1 = index / col2Vals.length;
+      int counter2 = index % col2Vals.length;
+      assert colVals[counter1]==Double.valueOf(cellVs[index][0].toString());
+      assert col2Vals[counter2]==Double.valueOf(cellVs[index][1].toString());
+      grab2DStats(tstats, fr, colVals[counter1], col2Vals[counter2], col, col2, cat, cat2, model);
+      // compare value from my manual to 2d pdp calculation
+      Assert.assertTrue(Math.abs(tstats[0]-Double.valueOf(cellVs[index][2].toString()))<tot);
```
[the remark this answers is not in the document]
Comment (wendycwong) [reply h2o-algos/src/test/java/hex/pdp/PartialDependenceTest.java]: Done.  Thanks!

File: h2o-core/src/main/java/hex/PartialDependence.java (hunk 1 above)
Comment (wendycwong) [reply h2o-core/src/main/java/hex/PartialDependence.java]: So picky!  Anyway, I tried to fix it.  Please check again and let me know.  Thanks, Wendy

File: h2o-core/src/main/java/hex/PartialDependence.java [hunk 4]
```diff
@@ -94,151 +121,189 @@ private void checkSanityAndFillParams() {
...
-                  stddevResponse[which] = Double.NaN;;
-                  stdErrorOfTheMeanResponse[which] = Double.NaN;;
+
+          int secondColLen = workingOn1D?1:col2Vals.length;
+          for (int j = 0; j < secondColLen; j++) {  // will go through at least once
+            final double value2 = workingOn1D?Double.NaN:col2Vals[j];
+            final int which = workingOn1D?k:j+k*col2Vals.length;
+            final boolean workOn2D = !workingOn1D;
+            
+            H2O.H2OCountedCompleter pdp = new H2O.H2OCountedCompleter() {
```
Comment (honzasterba) [inline h2o-core/src/main/java/hex/PartialDependence.java]: I think an anonymous inner class of this size deserves a name

File: h2o-py/h2o/model/model_base.py [hunk 5]
```diff
@@ -1054,51 +1076,129 @@ def partial_plot(self, data, cols, destination_key=None, nbins=20, weight_column
...
                 fig.tight_layout(pad=0.4, w_pad=0.5, h_pad=1.0)
             if(save_to_file is not None):
                 plt.savefig(save_to_file)
 
         return pps
-
+    
+    # change x, y, z to be 2-D numpy arrays in order to plot it.
+    # note that, x stays at one value for the duration of y value changes.
+    def predFor3D(self, x, y, z, colPairs, nbins, user_cols, user_num_splits):
```
Comment (navdeep-G) [inline h2o-py/h2o/model/model_base.py]: Is this a private function? If so, better to append with `__`

File: h2o-py/h2o/model/model_base.py [hunk 6]
```diff
@@ -1054,51 +1076,129 @@ def partial_plot(self, data, cols, destination_key=None, nbins=20, weight_column
...
+        if ycol in user_cols:
+            ind = user_cols.index(ycol)
+            nBins = user_num_splits[ind]
+        nrow = int(len(x)/nBins)
+        X = np.transpose(np.array(x).reshape(nrow, nBins))
+        Y = np.transpose(np.array(y).reshape(nrow, nBins))
+        Z = np.transpose(np.array(z).reshape(nrow, nBins))
+        return X,Y,Z
+    
+    def grabValues(self, pp, index, data, col, axs):
```
Comment (navdeep-G) [inline h2o-py/h2o/model/model_base.py]: Is this a private function? If so, better to append with `__`

File: h2o-py/h2o/model/model_base.py [hunk 7]
```diff
@@ -1054,51 +1076,129 @@ def partial_plot(self, data, cols, destination_key=None, nbins=20, weight_column
...
+            else:   # y-axis
+                axs.set_yticks(x)
+                axs.set_yticklabels(labels)
+            axs.margins(0.2) 
+
+            return xext
+        else:
+            return pp[index]
+        
+    def setAxs1D(self, axs, upper, lower, plot_stddev, cat, pp, pp_start_index, col):
```
Comment (navdeep-G) [inline h2o-py/h2o/model/model_base.py]: Is this a private function? If so, better to append with `__`

File: h2o-py/h2o/model/model_base.py [hunk 8]
```diff
@@ -1401,3 +1501,20 @@ def _get_matplotlib_pyplot(server):
     except ImportError:
         print("`matplotlib` library is required for this function!")
         return None
+
+def _get_mplot3d_pyplot(functionName):
```
Comment (navdeep-G) [inline h2o-py/h2o/model/model_base.py]: Typically private function in Python should start with `__`

File: h2o-py/h2o/model/model_base.py [hunk 9]
```diff
@@ -1401,3 +1501,20 @@ def _get_matplotlib_pyplot(server):
...
+    try:
+        # noinspection PyUnresolvedReferences
+        from mpl_toolkits.mplot3d import Axes3D
+        return Axes3D
+    except ImportError:
+        print("`mpl_toolkits.mplot3d` library is required for function {0}!".format(functionName))
+        return None
+
+def _get_numpy(functionName):
```
Comment (navdeep-G) [inline h2o-py/h2o/model/model_base.py]: Typically private function in Python should start with `__`

File: h2o-core/src/main/java/hex/PartialDependence.java (hunk 4 above)
Comment (wendycwong) [reply h2o-core/src/main/java/hex/PartialDependence.java]: I agree.  However, it seems like this is a normal practice here at H2O.  At least with the people who worked here before.

Comment (navdeep-G) [reply h2o-core/src/main/java/hex/PartialDependence.java]: I agree with @honzasterba. Better readability this way.

**Comment by wendycwong** [comment]
Honza:

I think it will be good to have support in Flow for this.  Let me open this one for you.

Regarding documentation, I made the change in the python/R client code already.  


Wendy

File: h2o-core/src/main/java/hex/PartialDependence.java (hunk 4 above)
Comment (honzasterba) [reply h2o-core/src/main/java/hex/PartialDependence.java]: there are many things that are frequent across the h2o codebase, but really should not be.
let's improve!

**Comment by honzasterba** [comment]
this part of docs might need updating:

http://docs.h2o.ai/h2o/latest-stable/h2o-docs/performance-and-prediction.html#partial-dependence-plots

File: h2o-py/h2o/model/model_base.py (hunk 5 above)
Comment (wendycwong) [reply h2o-py/h2o/model/model_base.py]: Yeap, fixed!  Thanks.

File: h2o-py/h2o/model/model_base.py (hunk 6 above)
Comment (wendycwong) [reply h2o-py/h2o/model/model_base.py]: Yeap, fixed! Thanks.

File: h2o-py/h2o/model/model_base.py (hunk 7 above)
Comment (wendycwong) [reply h2o-py/h2o/model/model_base.py]: Yeap, fixed! Thanks.

File: h2o-py/h2o/model/model_base.py (hunk 8 above)
Comment (wendycwong) [reply h2o-py/h2o/model/model_base.py]: Yeap, fixed! Thanks.

File: h2o-py/h2o/model/model_base.py (hunk 9 above)
Comment (wendycwong) [reply h2o-py/h2o/model/model_base.py]: Yeap, fixed! Thanks.

File: h2o-core/src/main/java/hex/PartialDependence.java (hunk 4 above)
Comment (navdeep-G) [reply h2o-core/src/main/java/hex/PartialDependence.java]: Be the change you want to see in this codebase @wendycwong ;)

**Comment by navdeep-G** [comment]
cc @angela0xdata @zainhaq-h2o Docs would need updating for this when the time comes.

File: h2o-core/src/main/java/hex/PartialDependence.java (hunk 4 above)
Comment (wendycwong) [reply h2o-core/src/main/java/hex/PartialDependence.java]: OMG, did you guys just finished watching some inspiring movies/shows or something?  Sure, I will take care of it.

**Comment by wendycwong** [comment]
Honza/Navdeep:

You guys are right about the documentation.  I will write a JIRA for that.

File: h2o-core/src/main/java/hex/PartialDependence.java (hunk 4 above)
Comment (wendycwong) [reply h2o-core/src/main/java/hex/PartialDependence.java]: Done moving pdp to its own glorious class.

**Comment by angela0xdata** [comment]
@navdeep-G @wendycwong @honzasterba if a doc change is required, please open a separate Jira ticket.

**Comment by wendycwong** [comment]
@angela0xdata @zainhaq-h2o Here is the doc update JIRA for 2D pdp: https://0xdata.atlassian.net/browse/PUBDEV-6598

File: h2o-algos/src/test/java/hex/pdp/PartialDependenceTest.java [hunk 10]
```diff
@@ -121,6 +118,127 @@ assert equalTwoDimTables(partialDependence._partial_dependence_data[1], partialD
...
+      partialDependence._user_splits = new double[]{65, 61, 72, 1.4, 6.7, 20};
+      partialDependence._frame_id = fr._key;
+      partialDependence.execImpl().get();
+      
+      Scope.track_generic(model);
+      Scope.track_generic(partialDependence);
+      Scope.track_generic(partialDependence1);
+      
+      // Compare 1d pdp results with and without 2D pdp.  They should be the same.
+      assert equalTwoDimTables(partialDependence._partial_dependence_data[0], partialDependence1._partial_dependence_data[0], 1e-10):
```
Comment (Pscheidl) [inline h2o-algos/src/test/java/hex/pdp/PartialDependenceTest.java]: There is `double tot = 1e-10;` defined at the very beginning (multiple occurances).

Comment (wendycwong) [reply h2o-algos/src/test/java/hex/pdp/PartialDependenceTest.java]: probably should define it at the top for everyone then.  I will put it in.  Thanks.

**Pull request merged**